File: backend/src/commands/loras.rs
use super::state::AppState;
use crate::blueprints::UninstallSummary;
use crate::download_manager::{self, DownloadSpec, EnsureOpts};
use crate::loras::{self, LoraPack, SaveUserLoraArgs};
use tauri::{AppHandle, State};

#[tauri::command]
#[specta::specta]
pub fn list_loras(app: AppHandle) -> Result<Vec<LoraPack>, String> {
    loras::list_loras(&app)
}

#[tauri::command]
#[specta::specta]
pub fn get_lora(app: AppHandle, id: String) -> Result<LoraPack, String> {
    loras::get_lora(&app, &id)
}

/// Enqueue LoRA variant install via Download Manager.
#[tauri::command]
#[specta::specta]
pub fn install_lora_variant(
    app: AppHandle,
    _state: State<'_, AppState>,
    id: String,
    arch: crate::recipe::RecipeArch,
) -> Result<(), String> {
    let _ = download_manager::ensure(
        &app,
        DownloadSpec::Lora { id, arch },
        EnsureOpts {
            wait: false,
            ..Default::default()
        },
    )?;
    Ok(())
}

/// Remove a LoRA variant weight file if unused by other ready variants.
#[tauri::command]
#[specta::specta]
pub fn uninstall_lora_variant(
    app: AppHandle,
    id: String,
    arch: crate::recipe::RecipeArch,
) -> Result<UninstallSummary, String> {
    loras::uninstall_variant(&app, &id, arch.as_str())
}

#[tauri::command]
#[specta::specta]
pub fn save_user_lora(app: AppHandle, args: SaveUserLoraArgs) -> Result<LoraPack, String> {
    loras::save_user_lora(&app, args)
}

#[tauri::command]
#[specta::specta]
pub fn delete_user_lora(app: AppHandle, id: String) -> Result<(), String> {
    loras::delete_user_lora(&app, &id)
}

#[tauri::command]
#[specta::specta]
pub fn open_user_loras_dir(app: AppHandle) -> Result<String, String> {
    loras::open_user_loras_dir(&app)
}

#[tauri::command]
#[specta::specta]
pub fn set_user_lora_thumbnail(
    app: AppHandle,
    id: String,
    bytes: Vec<u8>,
    ext: String,
) -> Result<String, String> {
    loras::set_user_lora_thumbnail(&app, &id, bytes, &ext)
}

#[tauri::command]
#[specta::specta]
pub fn clear_user_lora_thumbnail(app: AppHandle, id: String) -> Result<(), String> {
    loras::clear_user_lora_thumbnail(&app, &id)
}

File: frontend/lib/host/loras.ts
import { commands } from "@/lib/generated/bindings"
import type { RecipeArch } from "@/lib/arch"
import type {
  CivitaiLoraExpand,
  SaveUserLoraArgs,
  UninstallSummary,
} from "@/lib/generated/bindings"
import type { LoraPack } from "./types"

export type { UninstallSummary }

export type { CivitaiLoraExpand }

/** Expand a CivitAI URL into per-arch download URLs via `expand_civitai_lora_url`. */
export async function expandCivitaiLoraUrl(
  url: string
): Promise<CivitaiLoraExpand> {
  return commands.expandCivitaiLoraUrl(url)
}

/** Installed + user LoRA packs via `list_loras`. */
export async function listLoras(): Promise<LoraPack[]> {
  return commands.listLoras()
}

/** Single LoRA pack (variants, files, install state) via `get_lora`. */
export async function getLora(id: string): Promise<LoraPack> {
  return commands.getLora(id)
}

/** Enqueue a LoRA variant install via `install_lora_variant` (Download Manager). */
export async function installLoraVariant(
  id: string,
  arch: RecipeArch
): Promise<void> {
  await commands.installLoraVariant(id, arch)
}

/** Remove a variant weight if unused via `uninstall_lora_variant`. */
export async function uninstallLoraVariant(
  id: string,
  arch: RecipeArch
): Promise<UninstallSummary> {
  return commands.uninstallLoraVariant(id, arch)
}

/** Persist a user-authored LoRA pack via `save_user_lora`. */
export async function saveUserLora(input: SaveUserLoraArgs): Promise<LoraPack> {
  return commands.saveUserLora(input)
}

/** Delete a user-authored LoRA pack via `delete_user_lora`. */
export async function deleteUserLora(id: string): Promise<void> {
  await commands.deleteUserLora(id)
}

/** Open the user-LoRAs folder in the OS file manager via `open_user_loras_dir`. */
export async function openUserLorasDir(): Promise<string> {
  return commands.openUserLorasDir()
}

/** Write a custom thumbnail for a user LoRA via `set_user_lora_thumbnail`. */
export async function setUserLoraThumbnail(
  id: string,
  bytes: number[],
  ext: string
): Promise<string> {
  return commands.setUserLoraThumbnail(id, bytes, ext)
}

/** Drop a user LoRA's custom thumbnail via `clear_user_lora_thumbnail`. */
export async function clearUserLoraThumbnail(id: string): Promise<void> {
  await commands.clearUserLoraThumbnail(id)
}
